Implement `express_checkout` app

samitnuk opened the pull request:
#### PayPal Express

* Integrated [PayPal Server-Side SDK](https://developer.paypal.com/docs/checkout/reference/server-integration/setup-sdk/).
* Created `PaymentProcessor` class (as a wrapper for `paypal-checkout-serversdk`). Do we need to have it in `gateway.py`?
* Created `ExpressCheckoutTransaction` model to keep info about all tranactions related to one order.
* Added `mocked_data.py` with `dict`s that represent responses from PayPal API.



#### TODO:
- [ ] ~~Update docs~~.
- [x] Since we support only Django 2.2.x, should we use `path`s instead of `ulr`s.
- [x] I think we should not delete the `ExpressTransaction` model (because all related DB data will be deleted).
- [x] Add tests for `PAYPAL_BUYER_PAYS_ON_PAYPAL=True`. Manual tests confirm correct work. 
- [ ] Add tests and mocked data for `void_authorized_order` and `refund_order`. There were no tests for these cases in the initial implementation.
    * __NOTE__: mocked data for `refund_order` added (both `'minimal'` and `'representation'`). Order refunding is tested manually with `facade.refund_order` function.
- [ ] Do we need `ShippingOptionsView`? Since we do not allow to create redirect URL w/o address. Should we allow it?
Options we can set in "order create request body":
![image](https://user-images.githubusercontent.com/16734295/94665316-a658d180-0314-11eb-91b2-b3bfb11b1102.png)
    Currently, we set this as:
    ```
    'shipping_preference': 'SET_PROVIDED_ADDRESS' if address is not None else 'NO_SHIPPING'
    ```
    So `GET_FROM_FILE` is never used. Should we use it in some cases?

![image](https://user-images.githubusercontent.com/16734295/95344958-e3980300-08c2-11eb-8aa6-a4486971a3b2.png)

samitnuk committed:
Implement `express_checkout` app

samitnuk committed:
Add `testpaths` to `pytest.ini`

samitnuk committed:
Add test for views in `express_checkout` app

samitnuk committed:
Use `path`/`re_path` instead of `url`s

samitnuk committed:
Add test to check "order canceling" view

samitnuk committed:
Update `payment_details.html` template with new "PayPal button"

samitnuk committed:
Fix typos

sasha0 commented:
> In tests/unit/express/view_test.py not all tests are overridden. Some of them cannot be implemented with SDK???

Since we won't override anything for now, then shall we exclude this item from todo?

> Add tests for PAYPAL_BUYER_PAYS_ON_PAYPAL=True. Manual tests confirm correct work.
> Add tests and mocked data for void_authorized_order and refund_order. There were no tests for these cases in the initial implementation.

How much time that would approximately take?

solarissmoke committed:
Fix redirect URL name for paypal-payflow-detail in dashboard.

Also use Django's redirect shortcut for redirects for simplicity.

sasha0 commented:
> Update docs.

I'll address it separately by myself.

Commit:
Fix context processor settings for the sandbox.

Commit:
Adjust indents in the Paypal Express URLconfs.

Commit:
Adjust imports order with isort.

samitnuk committed:
Add test for `PAYPAL_BUYER_PAYS_ON_PAYPAL=True`

samitnuk committed:
Modify urls configuration to keep all tests for `paypal.express`

samitnuk committed:
Make `flake8` ignore `mocked_data.py`

Since we keep examples of responses here (as dicts), some line can be long and this is fine.

samitnuk committed:
Apply changes required by `flake8` and `isort`

samitnuk committed:
Fix URL in `payment_details.html` template
